### ISSUE TITLE
[Bug] Not failing when required fields are missing when using structured config

### ISSUE THREAD

**Issue by sotte** [facebookresearch/hydra#2193]
# 🐛 Bug
## Description
Hydra is not failing/throwing an error when required/mandatory fields (as specified by a structured config) are not filled. I would expect that not filled fields cause an error message. 

A not filled field clearly violates the schema which requires the field.

## Checklist
- [x] I checked on the latest version of Hydra
- [x] I created a minimal repro (See [this](https://stackoverflow.com/help/minimal-reproducible-example) for tips).

## To reproduce

This is the empty base config. Note the `foo` field is commented out, i.e. `foo` is not set.
```yaml
# conf/config.yaml
defaults:
  - base_config
  - _self_

# foo: test
```

Here we have a simple `BaseConfig` which only requires one field: `foo`.
```python
# htest2.py
from dataclasses import dataclass

import hydra
from hydra.core.config_store import ConfigStore
from omegaconf import MISSING, OmegaConf


@dataclass
class BaseConfig:
    foo: str = MISSING


cs = ConfigStore.instance()
cs.store(name="base_config", node=BaseConfig)


@hydra.main(config_path="conf", config_name="config")
def my_app(cfg: BaseConfig) -> None:
    print(OmegaConf.to_yaml(cfg))


if __name__ == "__main__":
    my_app()
```

When executing this, I would expect an error message telling me that the field `foo` is not set. However I just get a message:
```
python htest2.py
foo: ???
```

Is this the expected behavior?
If so can I enforce a check that all mandatory fields are set?
Note that you get an error when accessing the, well, missing field:
`omegaconf.errors.MissingMandatoryValue: Missing mandatory value: foo`. I would expect to get this right away when loading the config.

## System information
- **Hydra Version** :  1.1.2
- **Python version** :  3.9.7
- **Virtual environment type and version** :  venv
- **Operating system** :  linux

**Comment by Jasha10** [comment]
Hi @sotte,
Yes, the default behavior for Hydra is to pass the config to the main function even if there are missing keys values.

The latest [OmegaConf dev release](https://pypi.org/project/omegaconf/2.2.0.dev2/) installable from pypi includes an [`OmegaConf.missing_keys`](https://github.com/omry/omegaconf/blob/32b267d9b4e6c74957fa94122e8e15e0817d73e7/omegaconf/omegaconf.py#L815) function that can be used to check whether the given config has any missing keys. This functionality will be included in the upcoming OmegaConf 2.2 release.
```python
...

@hydra.main(...)
def my_app(cfg) -> None:
    missing_keys: set[str] = OmegaConf.missing_keys(cfg)
    if missing_keys:
        raise RuntimeError(f"Got missing keys in config:\n{missing_keys}")
    ...

**Comment by sotte** [comment]
Thanks @Jasha10!

So this is not a bug, but the expected behavior. Good to know.

I did not find this information in the docs. It it worth to mention that somewhere? A colleague was also surprised by this behavior.

Any idea when OmegaConf 2.2 is going to land and integrated into hydra? For now I would just turn the `missing_keys` into a local utility function.

**Comment by Jasha10** [comment]
We hope to release both OmegaConf 2.2 and Hydra 1.2 within one week!

> I did not find this information in the docs. It it worth to mention that somewhere? A colleague was also surprised by this behavior.

Yes, I think so.

**Comment by sotte** [comment]
> We hope to release both OmegaConf 2.2 and Hydra 1.2 within one week!

Nice!

I guess potential locations for mentioning this behavior would be here:

https://hydra.cc/docs/next/tutorials/structured_config/config_groups/#missing-fields
https://hydra.cc/docs/tutorials/structured_config/schema/

**Comment by oli-clive-griffin** [comment]
Hi all, I'm configuring hydra for a project at the moment and found this behaviour surprising too. Seeing as missing properties error when accessed, it seemed like making this checked at import / decorator time would be ideal. But I couldn't find an easy way to enforce this globally beyond using `OmegaConf.missing_keys` at the start of a `hyrda.main` function. Is there a better way to configure this?

**Comment by odelalleau** [comment]
> Is there a better way to configure this?

Not really, it's up to the user to check for missing keys and raise an exception if they want to.

**Comment by omry** [comment]
Thanks for raising this, and sorry for the long delay. This behavior is intentional. Hydra allows missing values to remain after composition so application code can populate them before access. Accessing a missing value before it is populated raises `MissingMandatoryValue`.

This is documented in https://hydra.cc/docs/tutorials/basic/your_first_app/using_config/, which describes `???` as a value that must be populated prior to access and demonstrates the access-time exception. Structured Configs document `MISSING` as the equivalent form at https://hydra.cc/docs/tutorials/structured_config/config_groups/#missing-fields.

Applications that want eager whole-config validation can opt in by checking `OmegaConf.missing_keys(cfg)`.
